xarray/core/computation.py

```diff
@@ -1855,12 +1855,19 @@ def where(cond, x, y, keep_attrs=None):
     Dataset.where, DataArray.where :
         equivalent methods
     """
+    from .dataarray import DataArray
+
     if keep_attrs is None:
         keep_attrs = _get_keep_attrs(default=False)
     if keep_attrs is True:
         # keep the attributes of x, the second parameter, by default to
         # be consistent with the `where` method of `DataArray` and `Dataset`
-        keep_attrs = lambda attrs, context: getattr(x, "attrs", {})
+        keep_attrs = lambda attrs, context: attrs[1]
+        # cast non-xarray objects to DataArray to get empty attrs
+        cond, x, y = (v if hasattr(v, "attrs") else DataArray(v) for v in [cond, x, y])
+        # explicitly broadcast to ensure we also get empty coord attrs
+        # take coord attrs preferentially from x, then y, then cond
+        x, y, cond = broadcast(x, y, cond)
```

Comment:
Sorry for the delay here @slevang 

It looks like `duck_array_ops.where` will make a copy of a broadcasted `x` or `y` if their dtypes are different:
https://github.com/pydata/xarray/blob/c86ffeac58fdc936633f42d6b2d6180e8791acd7/xarray/core/duck_array_ops.py#L293-L296

and 
https://github.com/pydata/xarray/blob/c86ffeac58fdc936633f42d6b2d6180e8791acd7/xarray/core/duck_array_ops.py#L194-L207

So this could possibly be a bad memory regression.

How about we explicitly copy over the attributes after the apply_ufunc call? The code will also be clearer

Reply:
Hmm, without the broadcast we fail on anything with two scalars e.g. `[DataArray, scalar, scalar]` because `len(attrs)` is 1 for the coord attrs list:
```python
  File "/home/slevang/xarray/xarray/core/merge.py", line 641, in merge_attrs
    return combine_attrs(variable_attrs, context=context)
  File "/home/slevang/xarray/xarray/core/computation.py", line 1865, in <lambda>
    keep_attrs = lambda attrs, context: attrs[1]
IndexError: list index out of range
```

Reply:
Right my suggestion is that we explicitly copy `attrs` over because there are many edge cases here

Reply:
I mean we fail within `apply_ufunc` if we pass the lambda as `keep_attrs`. So we're going to have to do something like "if user passes keep_attrs=True, then actually pass keep_attrs=False to apply_ufunc and rebuild the attrs later"

Reply:
Yes sounds good to me